src/replication.c

```diff
@@ -2386,6 +2482,303 @@ char *sendCommandArgv(connection *conn, int argc, char **argv, size_t *argv_lens
+        err = receiveSynchronousResponse(conn);
+        if (err == NULL) goto no_response_error;
+
+        if (err[0] == '-') {
+            serverLog(LL_NOTICE, "Server does not support sync with offset, RDB Channel Sync approach cannot be used: %s", err);
+            goto error;
+        }
+        sdsfree(err);
+        err = NULL;
+
+        if (connSyncWrite(conn, "SYNC\r\n",6, server.repl_syncio_timeout*1000) == -1) {
+            serverLog(LL_WARNING, "I/O error writing to Primary: %s",
+                connGetLastError(conn));
+            goto error;
+        }
+
+        server.repl_state = REPL_SEC_CONN_RECEIVE_ENDOFF;
+        return;
+    }
+    /* Receive ENDOFF reply */
+    if (server.repl_state == REPL_SEC_CONN_RECEIVE_ENDOFF) {
+        char buf[PROTO_IOBUF_LEN];
+        connSyncReadLine(conn, buf, 1024, server.repl_syncio_timeout * 1000);
+        if (buf[0] == '\0') {
+            /* Retry again later */
+            return;
+        }
+        if (processEndOffsetResponse(buf) == C_ERR) {
+            goto error;
+        }
+        if (reInitReplicaMainConnection(server.repl_transfer_s) == C_ERR) {
+            goto error;
+        }      
+        if (prepareRdbConnectionForRdbLoad(server.repl_full_sync_s) == C_ERR) {
+            goto error;
+        }
+        return;
+    }
+
+    no_response_error: /* Handle receiveSynchronousResponse() error when primary has no reply */
+        serverLog(LL_WARNING, "Primary did not respond to command during SYNC handshake");
+        /* Fall through to regular error handling */
+
+    error:
+        abortRdbConnectionSync(0);
+        return;
+
+    write_error: /* Handle sendCommand() errors. */
+        serverLog(LL_WARNING, "Sending command to primary in rdb channel replication handshake: %s", err);
+        sdsfree(err);
+        goto error;
+}
+
+void resizeReplicationBuffer(long long new_size) {
+    server.repl_data_buf->buf = zmremap(server.repl_data_buf->buf, server.repl_data_buf->size, new_size, MREMAP_MAYMOVE);
+    server.repl_data_buf->size = new_size;
+}
+
+void setReplicaBufferLimit() {
+    if (server.repl_data_buf && 
+        server.repl_data_buf->size > server.replica_full_sync_buffer_limit) {
+        /* Need to trim replication data buffer */
+        if (server.repl_data_buf->used > server.replica_full_sync_buffer_limit) {
+            /* Abort sync as the replica doesn't have enough space to store the replication buffer */
+            abortRdbConnectionSync(1);
+            return;
+        } else {
+            resizeReplicationBuffer(server.replica_full_sync_buffer_limit);
+        }
+    }
+}
+
+void replDataBufInit(connection *conn) {
+    client *c = connGetPrivateData(conn);
+    serverAssert(c->flags & CLIENT_MASTER);
+
+    server.repl_data_buf = zmalloc(sizeof(replDataBuf));
+    server.repl_data_buf->buf = zmmap(NULL, PROTO_IOBUF_LEN, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, 0, 0);
```

Comment:
There are several advantages to using mmap for allocating memory:
1. Anonymous mappings can reduce fragmentation
2. Unampping actually gives the memory back to the OS.
3. A single pointer is easier to use (compared to a linked list, we get a contiguous virtual memory), easier means less maintenance overhead.
4. MAP_HUGETLB can be used.

When mmap is used with exponential growth factor, less reallocations are required than when linear growth is used for linked lists.
For example, for linked list with 1024*16 as block size it will take around 6000 allocations to reach size of 1mb, comparing to ~30 using exponential grows factor.

Right now we free the buffer when we are done streaming it. We can consider shrinking it gradually, however the replica local streaming part is blocking so I am not sure freeing space is necessary. 

@xbasel

Reply:
i'm not sure these advantages are really right or justified.
1. that argument about reducing fragmentation can be said about anything else in redis. i.e. if each buffering subsystem will manage it's own memory directly to OS, and return it in the right order, rather than the allocator, it'll avoid fragmentation. on the other hand, the allocator let's you share things, so by not using the allocator, you're also NOT plugging fragmentation holes created by other subsystems. i don't think that's a reason to use mmap. besides, working with buffer chunks bigger than a page, isn't susceptible to fragmentation.
2. arguing that unmapping immediately gives the memory to the OS immediately, can be seen as a disadvantage, there's a reason why allocators don't immediately do it, it improves performance. if you think you can do better than jemalloc maybe starts a competing allcator :stuck_out_tongue: 
3. that's certainly true if you don't bother to dismiss its head :smile: 
4. i'd argue that this is a premature optimization. without showing any benchmarks or explicit reasoning, i'd go with the typical / simple approach.
5. not sure i understand the exponential growth thing. i suppose you can do the same with a greedy growth mechanism (allocating bigger blocks), and i suppose it can be counted as an advantage or a disadvantage.
6. the replica applies that buffer (calls processInputBuffer) in a blocking manner, but this loop consumes memory (processing commands that could be growing to the database), so releasing memory during that loop is desired.
7. the one advantage of a contiguous buffer that i can see, is that it would have been interesting to reduce the need of copying that data form that buffer to the query buffer, but i'm not sure it would be worth the complexity.

Reply:
I agree that not sharing memory and filling holes has its drawbacks. Also not releasing memory during the loop can be harmful. Identifying which method is the most effective will be challenging. 
I will move to linked list 👍